firefox_puppeteer/ui/tabbar.py

```diff
@@ -3,126 +3,315 @@
+         be a string with one of `menu`, `button` or `shortcut`, or a callback
+         which gets triggered with the current :class:`Tab` as parameter.
+         Defaults to `menu`.
+
+        :returns: :class:`Tab` instance for the opened tab
         """
-        if isinstance(target, int):
-            return self.tabs[target]
+        start_handles = self.marionette.window_handles
+
+        # Prepare action which triggers the opening of the browser window
+        if callable(trigger):
+            trigger(self.selected_tab)
+        elif trigger == 'button':
+            self.window.tabbar.newtab_button.click()
+        elif trigger == 'menu':
+            # TODO: Make use of menubar class once it supports ids
+            menu = self.window.marionette.find_element('id', 'menu_newNavigatorTab')
+            menu.click()
+        elif trigger == 'shortcut':
+            self.window.send_shortcut(self.window.get_localized_entity('tabCmd.commandkey'),
+                                      accel=True)
+        # elif - need to add other cases
+        else:
+            raise errors.InvalidValueError('Unknown opening method: "%s"' % trigger)
+
+        # TODO: Needs to be replaced with event handling code (bug 1121705)
+        Wait(self.marionette).until(
+            lambda mn: len(mn.window_handles) == len(start_handles) + 1)
+
+        handles = self.marionette.window_handles
+        [new_handle] = list(set(handles) - set(start_handles))
+        [new_tab] = [tab for tab in self.tabs if tab.handle == new_handle]
 
-        if isinstance(target, basestring):
+        # if the new tab is the currently selected tab, switch to it
+        if new_tab == self.selected_tab:
+            new_tab.switch_to()
+
+        return new_tab
+
+    def switch_to(self, target):
+        """Switches the context to the specified tab.
+
+        :param target: The tab to switch to. `target` can be an index, a :class:`Tab`
+        instance, or a callback that returns True in the context of the desired tab.
+
+        :returns: Instance of the selected :class:`Tab`
+        """
+        start_handle = self.marionette.current_window_handle
+
+        if isinstance(target, int):
+            return self.tabs[target].switch_to()
+        elif isinstance(target, Tab):
+            return target.switch_to()
+        if callable(target):
             for tab in self.tabs:
-                if target in tab.get_attribute('label'):
+                tab.switch_to()
+                if target(tab):
                     return tab
 
-            raise NoSuchElementException('Tab with a label containing "{}"" not'
-                                         ' found'.format(target))
+            self.marionette.switch_to_window(start_handle)
+            raise errors.UnknownTabError("No tab found for '{}'".format(target))
+
+        raise ValueError("The 'target' parameter must either be an index or a callable")
```

Comment:
Oh, afaics this came in from you. :) But that opens the question why I have created InvalidValueError, while we have an existing Python error here. I will remove our class and make use of ValueError instead.